src/mongo_odm/nvp.hpp

```diff
@@ -0,0 +1,202 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#define NVP(x) mongo_odm::makeNvp(&mongo_odm_wrap_base::x, #x)
+
+// Creates serialize() function
+#define ODM_SERIALIZE_KEYS                                                           \
+    template <class Archive>                                                         \
+    void serialize(Archive& ar) {                                                    \
+        serialize_recur<Archive, 0, std::tuple_size<decltype(fields())>::value>(ar); \
+    }                                                                                \
+    template <class Archive, size_t I, size_t N>                                     \
+    typename std::enable_if<(I < N), void>::type serialize_recur(Archive& ar) {      \
```

Comment:
`mongo_odm_serialize_impl`? Either way, if we're blatting this into the user's objects, let's add the prefix